Framework: allow the execution of code only once (#2048)
kuzzleio/kuzzle: doc/2/framework/classes/backend/install/index.md

---
code: true
type: page
title: install
description: Backend class install() method
---

# Install

<SinceBadge version="auto-version" />
<CustomBadge type="error" text="Experimental: non-backward compatible changes or removal may occur in any future release."/>

> Register code which will be executed only once during the lifetime of the app.
> 
> ::: info
> This method can only be used before the application is started because the handler is called when the app is launching. If the handler throws, the app will be unable to start.
> :::
> 

Reviewer: Wot? I don't see this as accurate as you may think. 
Why talk about the "lifetime" of the app if this is executed _pre-start_?

Instead of using a phrase and then making an _info_ block to explain it further, we could just simply explain it directly right?

```suggestion
Register setup code that runs only once and before the application starts.

::: info
If this method throws, the application won't start.
:::
```

Author: Your suggestion is quiet clear and it's far better but I feel like we still miss the "ever" part that I was trying to say by "lifetime of the app". When I read "that runs only once and before the application starts" I understand "each time I start the app my function will be executed" which is not the case... As long as you don't clear ES, the function will never run twice.

Author: Maybe we should add in the info block something like:
```suggestion
Register setup code that runs only once (as long as the database is preserved) and before the application starts.

::: info
If this method throws, the application won't start.
:::

```

Reviewer 2: It's not during the lifetime of an application. Rather, it's once during the lifetime of an app _environment_.
Also, I don't quite like the "as long as the database is preserved", as it introduces a technical detail into a feature description.

What about this?

```suggestion
Register code executed when the application starts, and only once on any given environment. Once successfully executed, the code associated to an install identifier will never be run again on that environment.

::: info
If this method throws, the application won't start.
:::
```
